### PR TITLE
Build AsyncResetSynchronizerShiftReg from new SynchronizerPrimitive

### PR DESCRIPTION
**Pull request by rmac-sifive**
**Type of change**: other enhancement

**Impact**: no functional change

**Development Phase**: implementation

**Release Notes**
`AsyncSynchronizerShiftReg` synchronizes W-bit wide signals with D stages of synchronizers and, previously, was constructed from D instantiations of a W-bit wide register. This PR changes that construction to instantiate W shift registers, each D stages deep, which is more friendly to industry CDC checking and tool flows

This PR also creates the module out of a new `SynchronizerPrimitive`, distinct from but functionally identical to `AsyncResetReg`, which makes it easier to substitute specific synchronizer cells during implementation

### CODE REVIEW

**Comment by rmac-sifive** [comment]
@jackkoenig understood that the conflict exists, but I'm not sure the answer is to remove `SynchronizerPrimitiveReg`. Implementation tools may want to use a different library cell for the synchronizer register vs a regular async reset register, and the module name will tell the tool which one to use

File: src/main/scala/util/ShiftReg.scala [hunk 1]
```diff
@@ -84,11 +85,41 @@ object AsyncResetShiftReg {
     apply (in, depth, init.litValue.toInt, None)
 }
 
+class SynchronizerPrimitiveShiftRegVec(depth: Int = 1, init: Int = 0) extends AbstractPipelineReg(1) {
```
Comment (mwachs5) [inline src/main/scala/util/ShiftReg.scala]: why is this a `Vec` if it is only 1 bit wide? I think yuo should drop the `Vec` from the name

File: src/main/scala/util/ShiftReg.scala [hunk 2]
```diff
@@ -84,11 +85,41 @@ object AsyncResetShiftReg {
...
+}
+
 // Note that it is important to override "name" in order to ensure that the Chisel dedup does
 // not try to merge instances of this with instances of the superclass.
-class AsyncResetSynchronizerShiftReg(w: Int = 1, sync: Int = 3, init: Int = 0) extends AsyncResetShiftReg(w, depth = sync, init, name = "sync") {
+class AsyncResetSynchronizerShiftReg(w: Int = 1, sync: Int = 3, init: Int = 0) extends AbstractPipelineReg(w) {
   require(sync > 0, "Sync must be greater than 0.")
   override def desiredName = s"AsyncResetSynchronizerShiftReg_w${w}_d${sync}_i${init}"
+
+  val output = Seq.tabulate(w) { i => SynchronizerPrimitiveShiftRegVec(io.d(i), depth = sync, init) }
```
Comment (mwachs5) [inline src/main/scala/util/ShiftReg.scala]: this looks good here... but I think we need to do something similar for the other `SynchronizerShiftReg` variants (or get rid of them)

Comment (rmac-sifive) [reply src/main/scala/util/ShiftReg.scala]: Sure, I changed `SyncResetSynchronizerShiftReg` to be constructed similarly. Basically WxD-sized reset synchronizers are created with W instantiations of D-deep 1-bit wide synchronizer shift regs. The other types are unchanged (constructed from D instances of a W-bit wide reg vec).

File: src/main/scala/util/ShiftReg.scala [hunk 3]
```diff
@@ -84,11 +87,41 @@ object AsyncResetShiftReg {
     apply (in, depth, init.litValue.toInt, None)
 }
 
+class SynchronizerPrimitiveShiftReg(depth: Int = 1, init: Int = 0) extends AbstractPipelineReg(1) {
```
Comment (mwachs5) [inline src/main/scala/util/ShiftReg.scala]: I think if this is primarily intended as a synchronizer primitive the default depth should be 3, or we should just remove the default and require the user to specify it.

File: src/main/scala/util/ShiftReg.scala [hunk 4]
```diff
@@ -136,7 +169,8 @@ class SyncResetSynchronizerShiftReg(w: Int = 1, sync: Int = 3, init: Int = 0) ex
 
   override def desiredName = s"SyncResetSynchronizerShiftReg_w${w}_d${sync}_i${init}"
 
-  io.q := ShiftRegInit(io.d, n = sync, init = init.U, name = Some("sync"))
+  val shiftRegs = List.tabulate(w) { i => ShiftRegInit[UInt](io.d(w), n = sync, init = init.U, name = Some(s"sync_${i}")) }
```
Comment (mwachs5) [inline src/main/scala/util/ShiftReg.scala]: but won't this not use the new Synchronizer primitive you've created? Maybe we should just delete this class entirely. I don't think we really need to make a sync-reset version of the synchronizer primitive. I'm not sure what in the code base is using this version.

The `SynchronizerShiftReg` is useful though, and should also be replaceable internally by your `SynchronizerPrimitive` but maybe should be renamed explicitly `NoResetSynchronizerShiftReg` and would just be a wrapper around the `AsyncResetSynchronizerShiftReg` that does a `withReset(false.B)`

File: src/main/scala/util/AsyncResetReg.scala [hunk 5]
```diff
@@ -48,6 +48,24 @@ class AsyncResetReg(resetValue: Int = 0) extends RawModule {
   io.q := reg
 }
 
+class SynchronizerPrimitiveReg(resetValue: Int = 0) extends RawModule {
```
Comment (mwachs5) [inline src/main/scala/util/AsyncResetReg.scala]: Let's just get rid of this

File: src/main/scala/util/ShiftReg.scala (hunk 3 above)
Comment (mwachs5) [reply src/main/scala/util/ShiftReg.scala]: Let's just build this with regular flops. And make sure it gets the proper inferred reset type.

**Comment by mwachs5** [comment]
@rmac-sifive this is superceded by https://github.com/chipsalliance/rocket-chip/pull/2212, so i am going to close this one.

**Pull request closed without merging**